### ISSUE TITLE
Debug assertions in `drop` cause panic-in-panic

### ISSUE THREAD

**Issue by pchickey** [stoklund/cretonne#192]
This may be a general rust knowledge problem, as its the first time I've dealt with this sort of issue -

I have some debug assertions of my own in my `FuncEnvironment` implementation that are going off as I work on my code. It seems that, sometimes, as the panic unwinds, they cause the debug assertions in the `drop` implementation to go off:
https://github.com/stoklund/cretonne/blob/master/lib/frontend/src/frontend.rs#L508

which leaves me with the rather unhelpful error message:

```
thread panicked while panicking. aborting.
error: An unknown error occurred
```

and, when RUST_BACKTRACE=1, erases the backtrace for my own error.

Is there a way to make those debug assertions in frontend.rs not go off if the thread is already panicking? Or is this just something I have to deal with?

**Comment by sunfishcode** [comment]
This has been problematic for me too, and it's the first time I've delt with this sort of issue too. In my experience though, with RUST_BACKTRACE=1, I get two backtraces, and the first one is the one I want, which is inconvenient but at least still usable.

I'm open to ideas here. I do like having assertions check that all blocks are sealed and filled, so my best idea so far is to introduce a new builder function like finalize() which could check such things. But then what if finalize() isn't called? ;-).

**Comment by bjorn3** [comment]
* Use @sunfishcode's finalize. Let it take self by value and let its drop panic, or
* Use https://doc.rust-lang.org/stable/std/thread/fn.panicking.html to not run the assertions while panicking

**Comment by stoklund** [comment]
Either would work, although I would prefer to avoid advanced runtime/language features for something as simple as this.

The assertions in a `Drop` impl essentially say "there are no circumstances where you would abort construction" which is a bit stark. If we add the `panicking()` checks, it becomes "there are no circumstances except for panicking where you would about construction". I don't think that it true either. What if I'm constructing a function on the fly using streaming data from a network connection. If the connection drops, I would likely abort construction of the current function, which seems like a completely reasonable thing to do.

I think a more idiomatic Rust pattern is to have a `finalize()` function which returns something you need to continue. That guarantees that the function gets called.

It is tempting to return the constructed `ir::Function` by value, but unfortunately that interferes with our goal of reusing memory allocations. This is a general problem with Rust's affine types: When you transfer ownership of an object by returning it, you also transfer the obligation to free memory used by the object. Which makes it hard to recycle allocations for multiple computations.

I haven't found bulletproof ways of using idiomatic affine type patterns without causing massive malloc churn. The `InstBuilder` interface is one attempt: Pass around by value a small object which contains references to the large objects with the allocations. I don't know if that method applies here.

**Comment by sunfishcode** [comment]
https://github.com/stoklund/cretonne/pull/193 is a PR that implements a `finalize()` function; see that PR for details.

**Comment by sunfishcode** [comment]
Fixed in #193.
